File: py/aoc2024/day4.py
"""
Day 4: Ceres Search
"""

SAMPLE_INPUT = """
MMMSXXMASM
MSAMXMSMSA
AMXSXMAAMM
MSAMASMSMX
XMASAMXAMM
XXAMMXXAMA
SMSMSASXSS
SAXAMASAAA
MAMMMXMMMM
MXMXAXMASX
"""


def part1(data: str) -> int:
    """
    >>> part1(SAMPLE_INPUT)
    18
    """
    lines = data.splitlines()
    return sum(
        all(
            (y2 := y + i * dy) in range(len(lines))
            and (x2 := x + i * dx) in range(len(lines[y2]))
            and lines[y2][x2] == c
            for i, c in enumerate("XMAS")
        )
        for y, line in enumerate(lines)
        for x in range(len(line))
        for dx in range(-1, 2)
        for dy in range(-1, 2)
    )


def part2(data: str) -> int:
    """
    >>> part2(SAMPLE_INPUT)
    9
    """
    lines = data.splitlines()
    return sum(
        "".join(chars) in ("MMASS", "MSAMS", "SMASM", "SSAMM")
        for above, line, below in zip(lines, lines[1:], lines[2:])
        for chars in zip(above, above[2:], line[1:], below, below[2:])
    )


parts = (part1, part2)

File: rs/src/day4.rs
use itertools::izip;

pub fn part1(data: &str) -> usize {
    let lines = &data.lines().collect::<Vec<_>>()[..];
    lines
        .iter()
        .enumerate()
        .flat_map(|(y, line)| {
            (0..line.len()).flat_map(move |x| {
                (-1..=1).flat_map(move |dy| {
                    (-1..=1).filter(move |dx| {
                        "XMAS".bytes().enumerate().all(|(i, b)| {
                            y.checked_add_signed(i as isize * dy)
                                .and_then(|y| lines.get(y))
                                .zip(x.checked_add_signed(i as isize * dx))
                                .and_then(|(line, x)| line.as_bytes().get(x))
                                == Some(&b)
                        })
                    })
                })
            })
        })
        .count()
}

pub fn part2(data: &str) -> usize {
    let lines = data.lines().collect::<Vec<_>>();
    izip!(&lines[..], &lines[1..], &lines[2..])
        .flat_map(|(above, line, below)| {
            izip!(
                above.bytes(),
                above.bytes().skip(2),
                line.bytes().skip(1),
                below.bytes(),
                below.bytes().skip(2)
            )
            .filter(|&(nw, ne, b, sw, se)| {
                b == b'A'
                    && (nw == b'M' && se == b'S' || se == b'M' && nw == b'S')
                    && (ne == b'M' && sw == b'S' || sw == b'M' && ne == b'S')
            })
        })
        .count()
}

#[cfg(test)]
mod tests {
    use super::*;
    use indoc::indoc;
    use pretty_assertions::assert_eq;

    static EXAMPLE: &str = indoc! {"
        MMMSXXMASM
        MSAMXMSMSA
        AMXSXMAAMM
        MSAMASMSMX
        XMASAMXAMM
        XXAMMXXAMA
        SMSMSASXSS
        SAXAMASAAA
        MAMMMXMMMM
        MXMXAXMASX
    "};

    #[test]
    fn part1_examples() {
        assert_eq!(18, part1(EXAMPLE));
    }

    #[test]
    fn part2_examples() {
        assert_eq!(9, part2(EXAMPLE));
    }
}
